### PR TITLE
fix: add condition render hook and forced tag selection

### PR DESCRIPTION
**Pull request by rebelchris**
### Describe what this PR does
- Added a new hook to conditionally evaluate experiments (to avoid allocaiton)
- Re-added onboarding 4.5 as `forced_tag_selection`

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-{number} #done

### Preview domain
https://as-209-forced-tag-selection-exp.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      return growthbook
+        ? growthbook.getFeatureValue(feature.id, feature.defaultValue)
+        : (feature.defaultValue as WidenPrimitives<T>);
+    },
+    {
+      enabled: shouldEvaluate,
+      ...disabledRefetch,
+    },
+  );
+
```
Comment (capJavert) [inline packages/shared/src/hooks/useConditionalFeature.ts:35]: We can probably return it from `initialData` instead of this `??` check?

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { disabledRefetch } from '../lib/func';
+
+export const useConditionalFeature = <T extends JSONValue>({
+  feature,
+  shouldEvaluate,
+}: {
+  feature: Feature<T>;
+  shouldEvaluate: boolean;
+}): WidenPrimitives<T> => {
+  const growthbook = useGrowthBook();
```
Comment (capJavert) [inline packages/shared/src/hooks/useConditionalFeature.ts:15]: I think it should come from `useGrowthBookContext`, not sure `useGrowthBook` gets the same instance?

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  const queryKey = generateQueryKey(
+    RequestKey.Feature,
+    null,
+    feature.id,
+    shouldEvaluate,
+  );
```
Comment (capJavert) [inline packages/shared/src/hooks/useConditionalFeature.ts:21]: Should it not be per user eg. for logouts? Also why is `shouldEvaluate`  needed, it is not used in queryFn so it just makes additional query key while disabled from what I can see.

**Review by capJavert — APPROVED** [review]
Looking good to me 👌

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  const { data: featureValue } = useQuery(
+    queryKey,
+    () => {
+      return growthbook
+        ? growthbook.getFeatureValue(feature.id, feature.defaultValue)
+        : (feature.defaultValue as WidenPrimitives<T>);
+    },
+    {
+      initialData: feature.defaultValue as WidenPrimitives<T>,
```
Comment (idoshamun) [inline packages/shared/src/hooks/useConditionalFeature.ts:29]: Shouldn't this be optional? Some cases would want to wait for the evaluation I guess

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  shouldEvaluate: boolean;
+}): WidenPrimitives<T> => {
+  const { user } = useContext(AuthContext);
+  const { growthbook } = useGrowthBookContext();
+  const queryKey = generateQueryKey(RequestKey.Feature, user, feature.id);
+
+  const { data: featureValue } = useQuery(
+    queryKey,
+    () => {
+      return growthbook
```
Comment (idoshamun) [inline packages/shared/src/hooks/useConditionalFeature.ts:24]: In case the gb object updates, shouldn't we retrigger this?

**Review by nensidosari — APPROVED** [review]
Looks good to me, glad to have an option to conditionally enroll

**Comment by nensidosari** [comment]
I'm thinking if there are going to be use cases where still this wouldn't be enough, but can't find any at the moment, so we should be good to go with this and then see if there is still any limitations.

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 5 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useConditionalFeature.ts:24]: I guess yes, not sure the current useFeatureValue does it.
But can add it in querykey somehow

File: packages/shared/src/hooks/useConditionalFeature.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  it('should return feature value if evaluation is true', async () => {
+    const { result } = renderHook(
+      () =>
+        useConditionalFeature({ feature: testFeature, shouldEvaluate: true }),
+      {
+        wrapper: Wrapper,
+      },
+    );
+
+    expect(result.current).toBe('default_value');
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/shared/src/hooks/useConditionalFeature.spec.tsx:59]: Yeah kind of missing the option to change enrollment in tests atm 🙈

File: packages/shared/src/hooks/useConditionalFeature.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+  it('should return feature value if evaluation is true', async () => {
+    const { result } = renderHook(
+      () =>
+        useConditionalFeature({
+          feature: testFeature,
+          shouldEvaluate: true,
+        }),
+      {
+        wrapper: createWrapper({ value: 'new_value' }),
```
Comment (rebelchris) [inline packages/shared/src/hooks/useConditionalFeature.spec.tsx:80]: @nensidosari added different test evaluation now :)

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 4 above)
Comment (capJavert) [reply packages/shared/src/hooks/useConditionalFeature.ts]: User is defaultValue (control) until proven different right? But loading state makes sense. Maybe we can also add an argument to hook for defaultValue so we can fallback it when needed?

File: packages/shared/src/hooks/useConditionalFeature.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  feature: Feature<T>;
+  shouldEvaluate: boolean;
+}): UseConditionalFeature<T> => {
+  const { user } = useContext(AuthContext);
+  const { growthbook } = useGrowthBookContext();
+  const queryKey = generateQueryKey(
+    RequestKey.Feature,
+    user,
+    feature.id,
+    growthbook,
```
Comment (capJavert) [inline packages/shared/src/hooks/useConditionalFeature.ts:27]: growthbook is a complex object I am not sure it would even do anything in query key since query key is stringified in react-query. Why is this needed?

Comment (capJavert) [reply packages/shared/src/hooks/useConditionalFeature.ts:27]: image

this is how query key looks 😆

Comment (rebelchris) [reply packages/shared/src/hooks/useConditionalFeature.ts:27]: Basically for [this](https://github.com/dailydotdev/apps/pull/2864#discussion_r1539387861), but not sure what alternatives we have?
There's not really a clear update/retrigger exposure I think.

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 5 above)
Comment (rebelchris) [reply packages/shared/src/hooks/useConditionalFeature.ts:33]: I'm actually not sure we need to retrigger?
Wouldn't it just reflect a next evaluation on it's own?

I think pretty much how everything works now (and why sometimes you need to hard refresh on apps)

Comment (idoshamun) [reply packages/shared/src/hooks/useConditionalFeature.ts:33]: ok so let's leave it as it is. but just double check that it works this way

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 8 above)
Comment (capJavert) [reply packages/shared/src/hooks/useConditionalFeature.ts:27]: Ah, our growthbook instance never changes currently during single page visit, and we don't add new features once initiated https://github.com/dailydotdev/apps/blob/2f31655db1810b3122ca7ca20c83b0350ca4aebc/packages/shared/src/components/GrowthBookProvider.tsx#L63-L87 so it should be good to go without looking at change.

That said if it was to ever change and we change logic in `GrowthBookProvider` better way would be to just invalidate `queryClient.invalidateQueries(RequestKey.Feature)` I think.

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 4 above)
Comment (idoshamun) [reply packages/shared/src/hooks/useConditionalFeature.ts]: ok, again we just have to watch out from flickering and layout shifts

File: packages/shared/src/hooks/useConditionalFeature.ts (hunk 8 above)
Comment (idoshamun) [reply packages/shared/src/hooks/useConditionalFeature.ts:27]: Good idea!

Comment (capJavert) [reply packages/shared/src/hooks/useConditionalFeature.ts:27]: Thats we have "you need to refresh" for experiment to become active thing on preview as well.

Comment (rebelchris) [reply packages/shared/src/hooks/useConditionalFeature.ts]: Moved we'll need to check on preview as locally doesn't really work to override.

**Review by idoshamun — APPROVED** [review]

**Comment by rebelchris** [comment]
@capJavert // @idoshamun let me know what you think.
Looks ok to me?

https://github.com/dailydotdev/apps/assets/554874/121dd715-9871-4439-ac32-da962623e76e

**Comment by idoshamun** [comment]
@rebelchris perfect!

**Review by capJavert — APPROVED** [review]
Looking good, if this works we should also add it to experimentation docs in storybook at some point since it is good utility 👌

**Comment by rebelchris** [comment]
@capJavert // @idoshamun See my test failed with not having the GB eval.
There's an edge case where it would be too fast, (gb not loaded)
So added the features ready state as enabled state + in the key.
Let me know if that works?
